drivers: fxos8700: enhancements, mma8652 compatibility, motion detection

jfischer-no opened the pull request:
drivers: fxos8700: enhancements, mma8652 compatibility, motion detection
* Add attr_set for SENSOR_ATTR_SAMPLING_FREQUENCY and allow to change output data rate
* Add power mode configuration
* Move whoami property to dts bindings
* Add basic support for motion detection

jfischer-no committed:
samples: fxos8700: add motion detection overlay and test

Add motion detection overlay and test.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
drivers: fxos8700: allow to change output data rate

Add attr_set for SENSOR_ATTR_SAMPLING_FREQUENCY and
allow to change output data rate.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
drivers: fxos8700: add motion detection

Add basic support for motion detection.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
drivers: fxos8700: add power mode configuration

Add power mode configuration.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
samples: fxos8700: remove DECIMATION_FACTOR

Remove DECIMATION_FACTOR from sample as the default
output data rate is set to 6.25Hz.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

galak commented on dts/bindings/sensor/nxp,fxos8700.yaml:
```diff
@@ -19,6 +19,12 @@ properties:
+      type: int
+      description: value of the WHOAMI register
+      generation: define
+      category: required
```
can we make it optional, since it was for early devices, and if this isn't set we assume the value is 0xc7

galak replied:
Can we still have it as optional, such that it the property is set, it means to set the WHO_AM_I register rather than just reading it.  It seems like the only reason we need this is for some early devices that didn't get WHO_AM_I set by the mfg.

jfischer-phytec-iot replied:
I have done it quite differently now

galak reviewed (changes requested):
We should make 'whoami' property for FXOS8700 optional in the device tree, if its not defined the default value of 0xc7 is assumed (we should make the comment about this in the yaml as binding description for the property and include the bits about pre-production devices that was in the Kconfig).

MaureenHelm commented on drivers/sensor/fxos8700/fxos8700.c:
```diff
@@ -358,6 +366,7 @@ static const struct sensor_driver_api fxos8700_driver_api = {
 	.channel_get = fxos8700_channel_get,
 #if CONFIG_FXOS8700_TRIGGER
 	.trigger_set = fxos8700_trigger_set,
+	.attr_set = fxos8700_attr_set,
```
Why is this inside the trigger conditional?

jfischer-phytec-iot replied:
fxos8700_attr_set os located in fxos8700_trigger.c and will be build only for trigger mode.
I also do not know for what it would be necessary in not trigger mode.

MaureenHelm commented on drivers/sensor/fxos8700/fxos8700_trigger.c:
```diff
@@ -136,6 +136,54 @@ static void fxos8700_work_handler(struct k_work *work)
+	struct fxos8700_data *data = dev->driver_data;
+
+	if (dr >= 800) {
+		dr = FXOS8700_CTRLREG1_DR_RATE_800;
```
This logic isn't correct when the sensor is in hybrid mode (accelerometer + magnetometer). The sample rates are halved for each encoding.

MaureenHelm commented on samples/sensor/fxos8700/prj.conf:
```diff
@@ -7,3 +7,10 @@ CONFIG_SYS_LOG_SENSOR_LEVEL=4
+
+CONFIG_FXOS8700_MODE_ACCEL=n
+CONFIG_FXOS8700_PM_LOW_POWER=n
+CONFIG_FXOS8700_DRDY_INT1=n
```
This will break boards that use INT1

jfischer-phytec-iot replied:
Why? ``frdm_k64f`` and ``hexiwear_k64`` use default INT2. I have inserted it to make easier to test the sample on ``frdm_kw41z`` and ``reel_board``, CONFIG_FXOS8700_DRDY_INT1 has to be changed manually to ``y``. See also  test-accel section in samples/sensor/fxos8700/sample.yaml

MaureenHelm commented on drivers/sensor/fxos8700/fxos8700.c:
```diff
@@ -311,6 +311,7 @@ static int fxos8700_init(struct device *dev)
 		return -EIO;
 	}
 
+
```
Stray edit

jfischer-no committed:
boards: frdm_kl25z: use fxos8700 driver for MMA8451

Use fxos8700 driver for MMA8451 accelerometer.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
drivers: fxos8700: add verification of the supported devices

Add basic verification of the supported devices.
Add IDs for MMA8451, MMA8652, MMA8653 and FXOS8700 and
remove Kconfig FXOS8700_WHOAMI option.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
samples: fxos8700: revised the output of the sensor values

Revised the output of the sensor values.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
boards: bbc_microbit: use fxos8700 driver for MMA8653

Use fxos8700 driver for MMA8653 accelerometer.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
Revert "samples: sensors: mma8451q: Add accelerometer sample"

Driver fxos8700 can also be used for the MMA8451
accelerometer and offers more functionality.
Revert the commit to avoid duplicate code.

This reverts commit 70a35e2346403ce9502283342b3d2b24fad125c7.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
Revert "drivers: sensor: Add support for MMA8451Q (3-axis accelerometer)"

Driver fxos8700 can also be used for the MMA8451
accelerometer and offers more functionality.
Revert the commit to avoid duplicate code.

This reverts commit 9c0d7813e55e323f6c2613faef0d6772fb016196.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
tests: build_all: remove MMA8451Q from dts.fixup

Remove MMA8451Q from dts.fixup.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-phytec-iot commented:
FYI @larsgk @WilliamGFish
Tested on frdm_k64f, frdm_kl25z, reel_board. I have no microbit here, it would be great if someone can test it. For accelerometer devices, please set CONFIG_FXOS8700_MODE_HYBRID=n and CONFIG_FXOS8700_MODE_ACCEL=y in samples/sensor/fxos8700/prj.conf

dbkinder reviewed (changes requested):
Looks like you're deleting a sample and its documentation,
and then adding a new sample (samples/sensor/fxos8700/src/main.c) without documentation? Can we get documentation for the new sample too?

WilliamGFish commented:
Johann,
Compiled and ran the sample for FXOS8700 on Micro:bit; **Successfully**

You could just create a file named prj_bbc_microbit.conf with the recommended config

galak reviewed (changes requested):
DTS's need a little work, we should keep names of the actual sensor, and use compatible to say it's like a fxos8700.

galak reviewed (dismissed):
We probalby need to cleanup / change the Kconfig to clarify which sensors this driver works for.  If I'm a user and have a board with a MMA8451Q how do tell to use the FX08700 driver?

jfischer-phytec-iot commented on boards/arm/frdm_kl25z/frdm_kl25z.dts:
```diff
@@ -64,11 +64,15 @@
+		compatible = "nxp,fxos8700";
 		reg = <0x1d>;
-		label = "MMA8451Q";
+		whoami = <0x1a>;
```
Yes, I forgot it then, will fix.

WilliamGFish commented:
@jfischer-phytec-iot 
It may sound silly but would the MAG3110 also be able to work under this driver? The FXOS has both accelerometer and magnetometer similar to the separate MMA8653 and MAG3110.

The micro:bit is based on the MMA and MAG so could assist with testing.

jfischer-phytec-iot commented:
>It may sound silly but would the MAG3110 also be able to work under this driver? The FXOS has both accelerometer and magnetometer similar to the separate MMA8653 and MAG3110.

I guess it is not that easy, register addresses are arranged differently and fxos8700 driver supports only one instance. I think it would be easier to write own driver for MAG3110. Btw, it may be not enough to write just a driver for MAG3110, please look at the AN4246.

MaureenHelm commented:
> > It may sound silly but would the MAG3110 also be able to work under this driver? The FXOS has both accelerometer and magnetometer similar to the separate MMA8653 and MAG3110.
> 
> I guess it is not that easy, register addresses are arranged differently and fxos8700 driver supports only one instance. I think it would be easier to write own driver for MAG3110. Btw, it may be not enough to write just a driver for MAG3110, please look at the AN4246.

It's not a silly question. :) I think I agree with @jfischer-phytec-iot about a separate mag3110 driver but need to investigate some more. On a side note, I have been considering enhancing the fxos8700 driver to support multiple instances.

MaureenHelm commented on drivers/sensor/fxos8700/fxos8700.c:
```diff
@@ -274,19 +273,32 @@ static int fxos8700_init(struct device *dev)
+	 */
 	if (i2c_reg_read_byte(data->i2c, config->i2c_address,
-			      FXOS8700_REG_WHOAMI, &whoami)) {
+			      FXOS8700_REG_WHOAMI, &data->whoami)) {
```
Why are you saving this into the data struct? Do you plan to use it again later?

jfischer-phytec-iot replied:
I have moved it here from fxos8700_config, it can be used to check if the device supports certain feature (e.g, in fxos8700_attr_set).

MaureenHelm commented on boards/arm/bbc_microbit/dts.fixup:
```diff
@@ -0,0 +1,5 @@
+#define CONFIG_FXOS8700_NAME			NORDIC_NRF_I2C_40003000_NXP_FXOS8700_1D_LABEL
```
Add a license/copyright header

jfischer-no committed:
boards: reel_board: use MMA8652FC label

Revise accelorometer node and use MMA8652FC label.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
drivers: fxos8700: mention compatible devices in the help section

Mention compatible devices in the Kconfig help section.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-phytec-iot commented:
>Looks like you're deleting a sample and its documentation,
and then adding a new sample (samples/sensor/fxos8700/src/main.c) without documentation? Can we get documentation for the new sample too?

@dbkinder Sample `samples/sensor/fxos8700` was already there. I have now inserted a simple README.rst, unfortunately I have no time for better documentation.

MaureenHelm commented on boards/arm/frdm_kl25z/Kconfig.board.orig:
```diff
@@ -0,0 +1,16 @@
+	depends on SOC_SERIES_KINETIS_KL2X
+	select SOC_PART_NUMBER_MKL25Z128VLK4
+<<<<<<< HEAD
+=======
```
Missed a conflict when rebasing

MaureenHelm commented on samples/sensor/fxos8700/src/main.c:
```diff
@@ -79,19 +77,28 @@ void main(void)
 		       sensor_value_to_double(&accel[1]),
-		       sensor_value_to_double(&accel[2]),
+		       sensor_value_to_double(&accel[2]));
+#ifndef CONFIG_FXOS8700_MODE_ACCEL
```
It would be more intuitive to do:
```c
#if defined(CONFIG_FXOS8700_MODE_MAGN) || defined(CONFIG_FXOS8700_MODE_HYBRID)
```

MaureenHelm commented on boards/arm/bbc_microbit/Kconfig.board:
```diff
@@ -7,3 +7,5 @@
 	bool "BBC MICRO:BIT"
 	depends on SOC_NRF51822_QFAA
+	select HAS_DTS_GPIO_DEVICE
+	select HAS_DTS_I2C_DEVICE if I2C
```
These two additions are no longer necessary. See #10309 and #10294

MaureenHelm commented on samples/sensor/fxos8700/README.rst:
```diff
@@ -1,38 +1,87 @@
 
-FXOS8700 Accelerometer/Magnetometer Sensor
-##########################################
+FXOS8700 Driver Sample
```
Please change the commit message. You're not adding a new README.rst, you're changing the existing one to also work with MMA8451, MMA8652, and MMA8653 accelerometers.

dbkinder commented on samples/sensor/fxos8700/README.rst:
```diff
@@ -1,38 +1,87 @@
-:ref:`hexiwear_k64` boards.  It does not work on QEMU.
+This project outputs sensor data to the console.
+For accelerometer devices like MMA8451Q, MMA8652FC or MMA8653FC,
+open the project configuration file :file:`prj.conf` set the following
```
s/set/and set/

dbkinder commented on samples/sensor/fxos8700/README.rst:
```diff
@@ -1,38 +1,87 @@
+
+Building and Running for Micro Bit
+==================================
+Micro Bitis is equipped with MMA8653FC accelerometer.
```
Change to:

   Micro Bit is equipped ...

jfischer-no committed:
samples: fxos8700: add accelerometer devices to README.rst

Add information about MMA8451Q, MMA8652FC and MMA8653FC
accelerometers to README.rst

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

MaureenHelm commented on samples/sensor/fxos8700/src/main.c:
```diff
@@ -8,6 +9,8 @@
 #include <sensor.h>
 #include <stdio.h>
 
+#define MOTION_THRESHOLD	17
```
This can be removed

MaureenHelm commented on samples/sensor/fxos8700/sample.yaml:
```diff
@@ -6,3 +6,13 @@ tests:
+      - CONFIG_FXOS8700_MODE_ACCEL=y
+      - CONFIG_FXOS8700_MOTION=y
+      - CONFIG_FXOS8700_MOTION_INT1=y
+      - CONFIG_FXOS8700_PM_LOW_POWER=y
```
Can we move these configs to a new `proj_motion.conf`?

jfischer-phytec-iot replied:
Done, but I have turned on motion and pulse detection in sample.yaml

jfischer-no committed:
samples: fxos8700: enable pulse and motion detection in sample.yaml

Enable pulse and motion detection in sample.yaml.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

MaureenHelm commented on samples/sensor/fxos8700/README.rst:
```diff
@@ -6,20 +6,70 @@ FXOS8700 Accelerometer/Magnetometer Sensor
+
+For accelerometer devices like MMA8451Q, MMA8652FC or MMA8653FC,
+open the project configuration file :file:`prj.conf` and set the following
+variables:
```
Let's provide a new `prj_accel.conf` file instead of asking users to manually edit the existing one.

WilliamGFish replied:
When you've finished this sample, I'll create something fit the micro: bit. 
I'll use the display to give some graphical output.

jfischer-phytec-iot replied:
done

jfischer-no committed:
samples: fxos8700: add config for accelerometer devices

Add configuration for accelerometer devices like
MMA8451, MMA8652 and MMA8653.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

MaureenHelm commented:
Please fix the merge conflicts
